[RenderSystem] Refactor for multi API support

blturkot opened the pull request:
- Removed multiple CDDI headers, Vk and DX12 uses single CDDI struct
- Moved common references to common.h
- Removed duplicates in Vulkan.cpp
- Fixed HeapMap accesses in CDDI
- Updated some CMake files to 3.5 minimal to get rid of deprecated warning
- Added CMakeSettings for folder view in Visual Studio

blturkot committed:
[RenderSystem] Major refactor for cross API support

Removed multiple CDDI headers and unify it.
Vulkan compiles ok, D3D12 have minor compile errors and ofc linker due to lack of function implementations.
Moved common references to common.h.
Removed duplicates in Vulkan.cpp.

blturkot committed:
Merge branch 'VkeDX12Port'

blturkot committed:
[RenderSystem] Minor fixes for compilation errors

Fixed HeapMap accesses in cross API.
Updates some CMake files to 3.5 minimal to get rid of deprecated warning.
Added CMakeSettings for folder view in Visual Studio.

przemyslaw-szymanski commented on .clang-format:
```diff
@@ -1,93 +1,103 @@
+# Indentation and spacing  
+BasedOnStyle: LLVM  
+Language: Cpp  
+Standard: c++17  
```
c++20

przemyslaw-szymanski commented on CMakeSettings.json:
```diff
@@ -0,0 +1,70 @@
+    },
+    {
+      "name": "VK-x64-Debug",
+      "generator": "Ninja",
```
is it working only with ninja? Would it work with default cmake generator?

blturkot replied:
Ninja is unified generator. Otherwise you have to be specific with version:
<img width="1080" height="73" alt="image" src="https://github.com/user-attachments/assets/a231826d-c3e1-445b-aed0-1a87ea60442b" />
I'd like to leave it with Ninja and developer is free to override that. BTW it will NOT affect solution generator, this is for Visual in folder view.

przemyslaw-szymanski commented on include/vke/RenderSystem/CDDI.h:
```diff
@@ -1,15 +1,236 @@
+        Result WaitForQueue(const NativeAPI::Queue &hQueue);
+        Result WaitForDevice();
+
+        NativeAPI::SImplementation &getInternal()
```
Get and there should be equivalent for const& GetInternal() const. Maybe change name to GetImplementation?

przemyslaw-szymanski commented on include/vke/RenderSystem/CDDI.h:
```diff
@@ -1,15 +1,236 @@
+        };
+
+    protected:
+        static DDIExtArray      svExtensions;
```
extensions and layers probably should go do SImplementation as well. DX12 does not have such thing rather.

przemyslaw-szymanski commented on include/vke/RenderSystem/Common.h:
```diff
@@ -109,41 +126,40 @@ namespace VKE
+
+        template <uint32_t DEFAULT_COUNT = 32> using UintVec = Utils::TCDynamicArray<uint32_t, DEFAULT_COUNT>;
+
+        VKE_DECLARE_HANDLE(Pipeline);
```
we would like to have spaces after all types of brackets.

blturkot replied:
Let me do this with next PR, I will implement clang-format with all requirements and then we'll run it on whole project.

przemyslaw-szymanski commented on include/vke/RenderSystem/Common.h:
```diff
@@ -185,198 +201,198 @@ namespace VKE
-                3 * 4, // B4G4R4A4_UNORM_PACK16,
-                5 + 6 + 5, // R5G6B5_UNORM_PACK16,
-                5 + 6 + 5, // B5G6R5_UNORM_PACK16,
+            static const uint16_t aSizes[] = {
```
open bracket in new line

blturkot replied:
As previously - I will do this in separate PR with clang-format changes.

przemyslaw-szymanski commented on include/vke/RenderSystem/Common.h:
```diff
@@ -910,52 +972,55 @@ namespace VKE
-                uint16_t                count;
-                BINDING_TYPE            type;
-                uint8_t                 binding;
+                const SamplerHandle     *ahSamplers;
```
& and * should be after type name not before variable name.

blturkot replied:
Planned for next PR.

przemyslaw-szymanski commented on include/vke/RenderSystem/Common.h:
```diff
@@ -2260,23 +2362,22 @@ namespace VKE
 
-        template<class T, uint32_t TASK_COUNT>
-        struct TaskPoolHelper
+        template <class T, uint32_t TASK_COUNT> struct TaskPoolHelper
```
template should be in line above

blturkot replied:
TBD in next PR.

przemyslaw-szymanski commented on include/vke/RenderSystem/Common.h:
```diff
@@ -2300,194 +2401,220 @@ namespace VKE
+                    }
+
+                    SVertexAttribute(DEFAULT_CTOR_INIT)
+                        : pName(""), format{ Formats::R32G32B32_SFLOAT }, vertexBufferBindingIndex{ 0 }, location{ 0 },
```
every parameter in initializer list should be in separate line starting with , like
format()
, vertexBuffer()
, location()

blturkot replied:
Also clang-format related.

przemyslaw-szymanski commented on src/RenderSystem/D3D12/CDDITypes.h:
```diff
@@ -253,37 +248,44 @@ namespace VKE
+            // TODO(blturkot): Fill with limits
+        };
+
+        using Buffer                = ID3D12Resource *;
```
should be no space before *

blturkot replied:
TBD

przemyslaw-szymanski commented on src/RenderSystem/Vulkan/CDDITypes.h:
```diff
@@ -1,57 +1,121 @@
-#   define VKE_USE_VULKAN_LINUX 1
-#   define VKE_USE_VULKAN_LINUX 1
-#   define VK_USE_PLATFORM_XCB_KHR 1
+#define VKE_USE_VULKAN_LINUX 1
```
indent was intentional here due to #if

blturkot replied:
clang-format messed with this. Added TBD for next PR.

blturkot committed:
[RenderSystem] Updated PR from feedback

- Updated standard for cpp20
- DDILayers and DDIExt are now in SImplementation
- Name and getter unification
